### DISCUSSION TITLE
New to OR tool [solving a puzzle]

### DISCUSSION THREAD

**Discussion by krakosaure** [google/or-tools#3042] [Linear Solver questions]
Hello all 

I am trying to use OR-Tools for the _first time_ to solve a `simple` puzzle. I am looking for all the 14 digits of a number so that:
  - every digit is between 1 and 9,
  - the number formed with the digits is either maximum or minimum
  - some strange condition is verified

I came up fairly quickly with a python script ... which gives me minimum and maximum when the logic implementing the `condition` is empty.  But I got errors:
  - (TypeError) when I try to multiply two values and 
  - (TypeError: unsupported operand type(s) for %: 'SumArray' and 'int') when I try to use "%"

All my variables are supposed to be integers.,

Did I miss something ?

If you want to have a look I have a gist [here](https://gist.github.com/krakosaure/433a749b0f916d087907e8022c62d748)

Thanks

**Comment by lperron** [comment]
As I said in the first post. Your problem is not linear (i.e. mul x y).

You should use CP-SAT for this. It supports `AddMultiplicationEquality()`.

Now I wonder what is the expected semantics. 

for instance:
```py
["add", "x", "-7"],
["eql", "x", "w"],
```
the solver does not support imperative programming. Any variable will have one fixed value in the solution.
So anything of the sort `x == x + 7` will automatically fail. Unless you have `x1 == x0 + 7`. But this requires parsing your program, detecting when you will need to create a new variable (in the sense of the OR model).

Furthermore:
```py
v = solver.IntVar(-infinity, infinity, f"v[{i}]")
print(f"v[{i}]: {inst}")
if inst[0] == "add":
    v = registers[op1] + op2
```
You are aware that you are creating a reference `v` to a solver object `solver.IntVar()`. Then you are simply overwriting this reference with `registers[op1] + op2`. It does not add any constraint.

So, all in all, you are completely missing the point with declarative programming.  And the semantic of you program is at best unclear in that context.

**Comment by Mizux** [comment]
seems it is a copy/paste/run tentative to solve https://adventofcode.com/2021/day/24 ;)

**Comment by krakosaure** [comment]
Thanks Laurent & Mizux for your answers.
Indeed, apologies I should may be have given more context. I decided to
give a shot to or-tools on day 24, it is a good way for me to learn the
tool.

Indeed I missed completely the point is this block:

    v = solver.IntVar(-infinity, infinity, f"v[{i}]")
    print(f"v[{i}]: {inst}")
    if inst[0] == "add":
        v = registers[op1] + op2

Indeed and the latest statement should probably become a constraint on
the new v variable as solver.Add(v == ...)
